### ISSUE TITLE
Signalr UnobservedTaskException

### ISSUE THREAD

**Issue by tnx247** [SignalR/SignalR#2960]
hello,
I handle `TaskScheduler.UnobservedTaskException` in `Application_Start`, and write to log by code:

```
TaskScheduler.UnobservedTaskException += (s, e) =>
            {
                e.SetObserved();
                e.Exception.Handle(t =>
                {
                    Logger.Error(t, "UnobservedTaskException");
                    return true;
                });
            };
```

Then when server in high load (about 3k connections opened), I sometimes receive logs, all of them are:

```
(>100 logs) System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'System.Web.WebSockets.AspNetWebSocket'.
   at System.Web.WebSockets.AspNetWebSocket.get_State()
   at Microsoft.AspNet.SignalR.WebSockets.WebSocketHandler.IsClosedOrClosedSent(WebSocket webSocket)
   at Microsoft.AspNet.SignalR.WebSockets.WebSocketHandler.<<CloseAsync>b__6>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNet.SignalR.Owin.OwinWebSocketHandler.<>c__DisplayClass1.<<RunWebSocketHandler>b__0>d__3.MoveNext()
```

```
(few logs) System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'System.Web.WebSockets.AspNetWebSocket'.
   at System.Web.WebSockets.AspNetWebSocket.get_State()
   at Microsoft.AspNet.SignalR.WebSockets.WebSocketHandler.<<SendAsync>b__0>d__2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNet.SignalR.Transports.WebSocketTransport.<PerformSend>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNet.SignalR.Transports.WebSocketTransport.<PerformSend>d__8.MoveNext()
```

```
(few logs) System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'System.Web.WebSockets.AspNetWebSocket'.
   at System.Web.WebSockets.AspNetWebSocket.get_State()
   at Microsoft.AspNet.SignalR.WebSockets.WebSocketHandler.SendAsync(ArraySegment`1 message, WebSocketMessageType messageType, Boolean endOfMessage)
   at Microsoft.AspNet.SignalR.WebSockets.DefaultWebSocketHandler.Flush()
   at Microsoft.AspNet.SignalR.Transports.WebSocketTransport.<PerformSend>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNet.SignalR.Transports.WebSocketTransport.<PerformSend>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNet.SignalR.Messaging.Subscription.<Invoke>d__2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNet.SignalR.Messaging.Subscription.<Work>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNet.SignalR.Messaging.MessageBroker.<DoWork>d__2.MoveNext()
```

```
(few logs) System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'System.Web.WebSockets.AspNetWebSocket'.
   at System.Web.WebSockets.AspNetWebSocket.get_State()
   at Microsoft.AspNet.SignalR.WebSockets.WebSocketHandler.SendAsync(ArraySegment`1 message, WebSocketMessageType messageType, Boolean endOfMessage)
   at Microsoft.AspNet.SignalR.WebSockets.DefaultWebSocketHandler.Flush()
   at Microsoft.AspNet.SignalR.Transports.WebSocketTransport.<PerformSend>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNet.SignalR.Transports.WebSocketTransport.<PerformSend>d__8.MoveNext()
```

Are they normal behavior? Should I need to care about them?

**Comment by DawtCom** [comment]
I have the same issue with my heavy hitter usage of signalR.  Is there any headway with this issue?

My StackTrace is:
```
An unobserved task exception has occurred. 
 System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. ---> System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'System.Web.WebSockets.AspNetWebSocket'.
   at System.Web.WebSockets.AspNetWebSocket.get_State()
   at Microsoft.AspNet.SignalR.WebSockets.WebSocketHandler.IsClosedOrClosedSent(WebSocket webSocket)
   at Microsoft.AspNet.SignalR.WebSockets.WebSocketHandler.<<CloseAsync>b__6>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNet.SignalR.Owin.OwinWebSocketHandler.<>c__DisplayClass1.<<RunWebSocketHandler>b__0>d__3.MoveNext()
   --- End of inner exception stack trace ---
---> (Inner Exception #0) System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'System.Web.WebSockets.AspNetWebSocket'.
   at System.Web.WebSockets.AspNetWebSocket.get_State()
   at Microsoft.AspNet.SignalR.WebSockets.WebSocketHandler.IsClosedOrClosedSent(WebSocket webSocket)
   at Microsoft.AspNet.SignalR.WebSockets.WebSocketHandler.<<CloseAsync>b__6>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNet.SignalR.Owin.OwinWebSocketHandler.<>c__DisplayClass1.<<RunWebSocketHandler>b__0>d__3.MoveNext()<---
```

**Comment by sdhebert** [comment]
In addition to the seeing the above stacktraces, another path involves AspNetWebSocket.get_State() via Messaging.Subscription:

```
Object name: 'System.Web.WebSockets.AspNetWebSocket'.
   at System.Web.WebSockets.AspNetWebSocket.get_State()
   at Microsoft.AspNet.SignalR.WebSockets.WebSocketHandler.<<SendAsync>b__0>d__2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNet.SignalR.Transports.WebSocketTransport.<PerformSend>d__8.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNet.SignalR.Messaging.Subscription.<Invoke>d__2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNet.SignalR.TaskAwaiterHelper.PreserveCultureAwaiter.GetResult()
   at Microsoft.AspNet.SignalR.Owin.OwinWebSocketHandler.<>c__DisplayClass1.<<RunWebSocketHandler>b__0>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNet.SignalR.Messaging.Subscription.<Work>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNet.SignalR.Messaging.MessageBroker.<DoWork>d__2.MoveNext()<---
```

**Comment by oferze** [comment]
@DamianEdwards @davidfowl @abnanda1 would you open and reprioritize this bug? happening to us as well (v2.2.0) and there are some comments by more people regarding the same problem.

**Comment by moozzyk** [comment]
Fixed with 80e519ad196527d8f85bbb2e1bdebfd0a07393b3 which is a merge commit for the proposed PR.

**Comment by stto3703** [comment]
Is version 2.2.1 going to be published to nuget anytime soon?
